### DISCUSSION TITLE
PETSC_OPTIONS

### DISCUSSION THREAD

**Discussion by LiuPengPeter** [idaholab/moose#16308] [Q&A Modules: General] [answered]
Hello everyone,
I meet the problem of converging.
In the Executioner, there maybe need PETSC options but I have no knowledge about this before.

For examlpe:
  petsc_options_iname = '-pc_type -sub_pc_type   -sub_pc_factor_shift_type'
  petsc_options_value = 'asm       ilu            nonzero'

Is there related material provided for beginners?

Thank you a lot!

**Comment by fdkong** [comment]
PETSc manual is a great doc you could look at https://www.mcs.anl.gov/petsc/petsc-current/docs/manual.pdf

For these options, you could find them on petsc website.  For example, 


`-pc_type` https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/PC/PCSetType.html#PCSetType

`asm` https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/PC/PCASM.html#PCASM

`ilu` https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/PC/PCILU.html#PCILU

**Comment by YaqiWang** [comment, marked as answer]
Some basic explanations on MOOSE executioners can be found here: https://mooseframework.inl.gov/source/executioners/Steady.html. It should give enough information to users on how MOOSE interact with PETSc.

**Reply by LiuPengPeter** [reply]
Thank you guys. It seems so difficult and needs much time to understand the setting of executioners.
Is it possible I copy from other guys, like this:
```
[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = 'PJFNK'
  #########PETSC_OPTIONS###################very_important###################
  petsc_options_iname = '-pc_type -sub_pc_type   -sub_pc_factor_shift_type'
  petsc_options_value = 'asm       ilu            nonzero'
  #solve_type =Newton
  l_max_its = 50
  l_tol = 1e-4
  nl_max_its = 20
  nl_rel_tol = 1e-10
  nl_abs_tol = 1e-10
  end_time = 2.00E+11
[./TimeStepper]
    ## Turn on time stepping
    type = IterationAdaptiveDT
    dt = 1.00E+04 #06
    cutback_factor = 0.8
    growth_factor = 2.0 #1.5
    optimal_iterations = 7
 [../]
[]
```

**Reply by YaqiWang** [reply]
It is especially when you are solving a similar problem as the one you copied from but I still suggest you to understand what each parameter is doing. The settings for one problem may not be the best for another. Some of these parameters are pretty strait-forward to users like `type`, `end_time`, `dt`.  Some are intermediate like `l_max_its`, `l_tol`, `nl_max_its`, `nl_rel_tol`, `nl_abs_tol`, `optimal_iterations`, `cutback_factor`, `growth_factor`. Some are difficult like `scheme`, `solve_type`, `petsc_options_iname`, `petsc_options_value`. It is worthwhile to spend some time on the parameter documentations.

**Reply by LiuPengPeter** [reply]
Thank  you a lot!
